### PR TITLE
Adding tests for `kedge build`

### PR DESCRIPTION
**Pull request by ashetty1**
The test here creates a local docker registry, runs `kedge build`, and verifies if the image has been uploaded.

### CODE REVIEW

**Comment by ashetty1** [comment]
@surajssd would like to have your inputs on this.

**Comment by ashetty1** [comment]
Working on fixing the travis errors

File: tests/e2e/build_test.go [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+	// Delete the registry
+	_, err = runCmd("docker stop kedge_registry")
+	if err != nil {
+		t.Error(err)
+	}
+
+	_, err = runCmd("docker rm -v kedge_registry")
+	if err != nil {
+		t.Error(err)
+	}
```
Comment (surajssd) [inline tests/e2e/build_test.go]: put all of the clean up code in a fucntion and then call it with defer so that we always do the clean up!

**Comment by surajssd** [comment]
@ashetty1 tried running it, but getting following error, I think this should also be taken care!

```console
$ go test -run TestBuild github.com/kedgeproject/kedge/tests/e2e
--- FAIL: TestBuild (0.01s)
panic: error running command &{/bin/sh [/bin/sh -c docker run -d -p 5000:5000 --restart=always --name kedge_registry registry:2] []  <nil>  docker: Error response from daemon: Conflict. The container name "/kedge_registry" is already in use by container "889c30df6d0c704c862790f8a6ef489c5e9a3b93638aadf57ce301d402e914d9". You have to remove (or rename) that container to be able to reuse that name.
See 'docker run --help'.
 [] <nil> 0xc4201a0930 exit status 125 <nil> <nil> true [0xc420302098 0xc4203020b8 0xc4203020d0] [0xc420302098 0xc4203020b8 0xc4203020d0] [0xc4203020a8 0xc4203020c8] [0x552540 0x552540] 0xc4201ce660 <nil>}: exit status 125 [recovered]
        panic: error running command &{/bin/sh [/bin/sh -c docker run -d -p 5000:5000 --restart=always --name kedge_registry registry:2] []  <nil>  docker: Error response from daemon: Conflict. The container name "/kedge_registry" is already in use by container "889c30df6d0c704c862790f8a6ef489c5e9a3b93638aadf57ce301d402e914d9". You have to remove (or rename) that container to be able to reuse that name.
See 'docker run --help'.
 [] <nil> 0xc4201a0930 exit status 125 <nil> <nil> true [0xc420302098 0xc4203020b8 0xc4203020d0] [0xc420302098 0xc4203020b8 0xc4203020d0] [0xc4203020a8 0xc4203020c8] [0x552540 0x552540] 0xc4201ce660 <nil>}: exit status 125

goroutine 20 [running]:
testing.tRunner.func1(0xc4202b60f0)
        /usr/local/go/src/testing/testing.go:711 +0x2d2
panic(0xfd1000, 0xc42044a080)
        /usr/local/go/src/runtime/panic.go:491 +0x283
github.com/kedgeproject/kedge/tests/e2e.TestBuild(0xc4202b60f0)
        /home/fedora/go/src/github.com/kedgeproject/kedge/tests/e2e/build_test.go:60 +0x80e
testing.tRunner(0xc4202b60f0, 0x11c3820)
        /usr/local/go/src/testing/testing.go:746 +0xd0
created by testing.(*T).Run
        /usr/local/go/src/testing/testing.go:789 +0x2de
FAIL    github.com/kedgeproject/kedge/tests/e2e 0.068s
```

**Comment by surajssd** [comment]
After cleaning up manually

```console
$ docker stop kedge_registry
$ docker rm -f $(docker ps -aq)
```

I ran that individual test and it took almost 4mins to complete:

```console
$ go test -run TestBuild github.com/kedgeproject/kedge/tests/e2e                                                                                                                        
ok      github.com/kedgeproject/kedge/tests/e2e 231.089s
```

This should run in parallel with e2e tests, to make it faster.

**Comment by surajssd** [comment]
Also it is using the docker env of the host, not of the minikube or minishift vm!

**Comment by ashetty1** [comment]
@surajssd I am not sure if we need to check if there is a docker registry with the same name before running the test. What we should expect before the test is a clean setup and then leave the system as it was once the test run is complete

File: tests/e2e/build_test.go [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+	// Context Dir to be passed for kedge build
+	var contextDir = KedgeLoc + "/docs/examples/build"
+
+	endpoint := "unix:///var/run/docker.sock"
+	cli, err := docker.NewClient(endpoint)
+	if err != nil {
+		t.Error(err)
+	}
+
+	dockerRegStart := "docker run -d -p 5000:5000 --restart=always --name kedge_registry registry:2"
```
Comment (surajssd) [inline tests/e2e/build_test.go]: You can also add the flag `--rm` which will delete the docker container once the registry is stopped!

Comment (surajssd) [reply tests/e2e/build_test.go]: This will reduce the whole hassle of deleting the container

**Comment by surajssd** [comment]
>@surajssd I am not sure if we need to check if there is a docker registry with the same name before running the test. What we should expect before the test is a clean setup and then leave the system as it was once the test run is complete

why should we assume that? We will also be running tests locally so we don't want to leave the environment dirty in dev machine!

**Comment by ashetty1** [comment]
@surajssd kindly review this.

File: tests/e2e/build_test.go [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+	reg, err := runCmd(dockerRegStart)
+	if err != nil {
+		panic(err)
+	}
+
+	t.Logf("Local docker registry has been set up: %s", reg)
+
+	defer DeleteRegistry(t)
+
+	runKedgeCmd := KedgeLoc + "/kedge build -c " + contextDir + " -i localhost:5000/test:2.0 -p"
```
Comment (cdrage) [inline tests/e2e/build_test.go]: have a look at https://github.com/kedgeproject/kedge/blob/master/tests/e2e/e2e_test.go#L30

would be better to throw the binary location up in global variables

File: tests/e2e/build_test.go [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+	t.Log("The registry has been deleted successfully")
+}
+
+func TestBuild(t *testing.T) {
+	// Context Dir to be passed for kedge build
+	var contextDir = KedgeLoc + "/docs/examples/build"
+
+	// Cleanup
+	_, _ = runCmd("docker stop kedge_registry")
```
Comment (cdrage) [inline tests/e2e/build_test.go]: consider adding a slower timeout `-t 1` in order to speed up tests.

File: tests/e2e/build_test.go (hunk 2 above)
Comment (cdrage) [reply tests/e2e/build_test.go]: :+1:

**Comment by ashetty1** [comment]
@cdrage could you please do a final review here.

**Comment by cdrage** [comment]
This LGTM :+1: merge when you're comfortable.

**Comment by ashetty1** [comment]
@surajssd @kadel let me know if you guys think this can be merged.

**Comment by kadel** [comment]
this already has two LGTMs.
I'm merging this

**Pull request merged**